Bluetooth: controller: split: non-conn non-scan advertising with aux packets

cvinayak opened the pull request:
Add implementation to support ADV_EXT_IND PDU with AuxPtr. Changes in this commits prepares the ADV_EXT_IND PDU in the primary channel, prepares the AUX_ADV_IND PDU in the auxiliary channel.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Add Adv. Ext. HCI debug informations

Added debug information print outs in HCI code related to
Advertising Extensions events.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: shell: Updated advx command to use Adv Data Set

Updated the shell commands related to Advertising Extensions
to call the LE Set Extended Advertising Data command
functional interface for testing purposes.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Added le_set_ext_adv_param interface

Added le_set_ext_adv_param functional interface to integrate
HCI with controller's functional interface.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Add AuxPtr to ADV_EXT_IND PDU

Add implementation to support ADV_EXT_IND PDU with AuxPtr.
Changes in this commit only prepares the ADV_EXT_IND PDU
in the primary channel.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Move Adv Ext interface to ll interface

Move Advertising Extensions interface to LL interface header
file.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: non-conn non-scan with aux packets

Added implementation to advertise non-connectable
non-scannable advertising with auxiliary packets.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: advx: test application for adv. ext. development

Added a new advx test application to aid test driven
development.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Minor refactor of SyncInfo structure

Minor refactor of SyncInfo structure field names.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fill bdaddr in aux packet

Fill the bdaddr on advertising enable in the aux packet.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Refactor to reuse common ISR code

Refactor to reuse common repeated code in ISR that reset
status, and performed abort, done and cleanup.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Move access address helper to util.c

Move the access address helper function to util.c file.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Moved chan map helper functions

Moved Channel Map helper functions into separate file.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Use correct secondary PHY, AA and CRC

Implement use of correct secondary PHY for transmission,
access address and CRC setup.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

aescolar commented:
Added 2.3 milestone to make it clear it is  not for 2.2 and get it out of filters output

cvinayak committed:
Bluetooth: controller: split: Added SyncInfo into Aux Packet

Added SyncInfo population in Aux packet.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Handle SyncInfo in Aux Data set

Handle using existing SyncInfo when updating Aux PDU AD Data.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: advx: Update data while periodic advertising

Update Aux PDU AD Data while periodic advertising.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix to use correct secondary PHY

Fix using correct secondary PHY for auxiliary PDUs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: advx: Use 2M PHY auxiliary PDUs.

Use 2M PHY auxiliary PDUs in test.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Implement Aux Sets

Implement use of configurable Auxiliary advertising sets.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

kruithofa commented:
I need some more time to look through the changes here

kruithofa commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1616,6 +1616,39 @@ static void le_enh_tx_test(struct net_buf *buf, struct net_buf **evt)
+	tx_pwr = cmd->tx_power;
+
+	if (!(evt_prop & BIT(4))) {
+		adv_type = 0x05; /* Extending advertising */
```
Use a #define instead of 0x05

cvinayak replied:
addressed.

kruithofa commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1616,6 +1616,39 @@ static void le_enh_tx_test(struct net_buf *buf, struct net_buf **evt)
+				   cmd->sid, cmd->scan_req_notify_enable);
+
+	rp = hci_cmd_complete(evt, sizeof(*rp));
+	rp->status = 0x00;
```
use a define here as well?

cvinayak replied:
success is generally returned as 0.

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -37,36 +37,173 @@ u8_t *ll_adv_aux_random_addr_get(u8_t handle, u8_t *addr)
+	 */
+
+	/* TODO: handle other op values */
+	if ((op != 0x03) && (op != 0x04)) {
```
use #define for constants

cvinayak replied:
addressed.

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -214,7 +213,7 @@ u8_t ll_adv_params_set(u16_t interval, u8_t adv_type,
 
 		/* Zero-init header flags */
-		if (is_pdu_type_changed) {
+		if (is_pdu_type_changed || (adv_type == 0x06)) {
```
use define for adv_type

cvinayak replied:
addressed/stale.

kruithofa reviewed (commented):
This seems more a sample instead of a test; consider moving to sample/bluetooth directory
missing .yaml file?

cvinayak committed:
tests: Bluetooth: advx: disable periodic SYNC_IND packets

Added function call to test disable of the started periodic
SYNC_IND packets.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Add periodic adv LLL implementation

Add LLL implementation to perform periodic advertising.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Add periodic adv ULL implementation

Add ULL implementation to perform periodic advertising.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: advx: Added code to test periodic advertising

Added code to invoke starting of periodic advertising.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix piggy back implementation

Fix conditional compilation of piggy back implementation.
This code should be deleted once periodic advertising
implementation is complete.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix cond. complile of periodic adv

Fix conditional compilation of advertising extensions and
periodic advertising do that platforms with not support in
their LLL can be built without advertising extensions.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Port openisa files

Port OpenISA LLL files to reflect some of the changes done
in Nordic LLL.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: handle latency for cancelled conn events

Update implementation of master and slave LLL's to correctly
handle event counter values when latencies introduced due to
connection events cancelled by active events operating in
unreserved time space.

When an active radio event extends into unreserved time
space, and a connection event prepare is scheduled but at
the time of pre-emption timeout if the connection event is
cancelled then the event count and latencies needs to be
continiued to get acummulated.

In the current controller usecases the above scenarios does
not get exercised, the changes in this commit is needed for
future roles that can extend into unreserved time space and
would cancel a scheduled connection event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: handle latency for cancelled sync events

Update implementation of periodic advertiser's LLL to
correctly handle event counter values when latencies
introduced due to sync events cancelled by active events
operating in unreserved time space.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix uninitialized variable use error

Fix for compiler complaining of uninitialized variable use.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: cond compile Ext Adv without Aux chan

Conditional compile Extended Advertising without Auxiliary
channel PDU use.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Enable Adv Extensions

Enable Advertising Extensions features in the controller
when enabled in the Host in combined host+controller based
applications.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: hci: Integrate Adv Extensions commands

Integrate Advertising Extensions related Extended Advertising
and Extended Scan HCI commands.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: hci: Indent BT_INFO of extended header

Indent when logging out the field value of common extended
advertising header format.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Use uint8_t for scan handles

Use uint8_t for scan set handles.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: use CONFIG_BT_CTLR_ADV_SET

Use CONFIG_BT_CTLR_ADV_SET instead of derived BT_CTLR_ADV_SET.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Differentiate extended ll_adv_params_set

Differentiate call to ll_adv_params_set, whether used by
legacy LE Set Advertising Parameters or LE Set Extended
Advertising Parameters.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Support legacy AD/SR data

Added implementation to support setting of legacy
Advertising Data and Scan Response Data using Extended
Advertising Data set functions.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Added Adv Random Address set

Added implementation to set advertising random address.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Added Extended Scan Support

Updated implementation to support Extended Scan Parameters
and Extended Scan Enable.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix missing reset of scanner state

Fix missing reset of scanner state which could cause a
spurious SCAN_RSP packet without a prior SCAN_REQ being sent
to be reported.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Add the missing enable of dup filtering

Add implementation to enable duplicate filtering for extended
scanning.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: SCAN_RSP for ADV_IND or SCAN_IND

Added implementation to differentiate SCAN_RSP received for
ADV_IND or SCAN_IND PDU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Add generate of ext adv legacy report

Added implementation to generate LE Extended Advertising
Report for received legacy PDUs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Added LE Ext. Create Connection

Added implementation of LE Extended Create Connection.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: hci: Integrate LE Ext. Create Connection

Integrate the support for HCI LE Extended Create Connection.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix missing coded PHY scanning

Fix, use phy variable instead of type to determine if
scanning on Coded PHY is to be enabled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Allow set rand addr when using ext cmds

Allow use of HCI LE Set Random Address cmd when advertising
is enabled using Extended Advertising commands.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix extended scan param set

Fix missing reset of phy struct member for scan set that is
not being enabled by use of LE Extended Scan Parameter Set
command.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Workaround succeed on zero len scan rsp set

Added a workaround to succeed setting a zero length Extended
Scan Response Data.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_internal.h:
```diff
@@ -1,23 +1,64 @@
+struct ll_adv_set *ull_adv_is_created_get(u8_t handle);
+
+/* Helper function to construct AD data */
+u8_t adv_data_set(struct ll_adv_set *adv, u8_t len, u8_t const *const data);
```
I think this function name should have `ull_` as prefix.

cvinayak replied:
addressed.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_internal.h:
```diff
@@ -1,23 +1,64 @@
+u8_t adv_data_set(struct ll_adv_set *adv, u8_t len, u8_t const *const data);
+
+/* Helper function to construct SR data */
+u8_t scan_rsp_set(struct ll_adv_set *adv, u8_t len, u8_t const *const data);
```
I think this function name should have `ull_` as prefix.

cvinayak replied:
addressed.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -104,3 +512,259 @@ u8_t ll_adv_aux_set_clear(void)
+			     sizeof(struct ll_adv_aux_set));
+}
+
+static void mfy_aux_offset_get(void *param)
```
As far as I understand, this function determines where to place the auxiliary advertisement. It seems to be quite elaborate for what it does. Can you maybe explain why this is the preferred way of doing this?

cvinayak replied:
no action.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -104,3 +512,259 @@ u8_t ll_adv_aux_set_clear(void)
+
+		sync  = (void *)HDR_LLL2EVT(adv->lll.sync);
+		if (sync->is_started) {
+			ull_adv_sync_offset_get(adv);
```
Doesn't calling this function after the prepare-mayfly has already been enqueued cause a race condition?

cvinayak replied:
Not in nRF, as the PDU is picked at start/anchor. But yes, if the PDU is not populated in time, then its an issue. I think this be addressed in separate GH issue/PR to synchronize PDU population during prepare timespace.

cvinayak replied:
Create a GH issue for future

cvinayak replied:
Create a GH issue for future

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_master.c:
```diff
@@ -51,32 +52,67 @@
+
 	if (lll->conn) {
-		return BT_HCI_ERR_CMD_DISALLOWED;
+		goto conn_is_valid;
```
This seems to change semantics for the case that `lll->conn!=0` without advertisement extensions enabled (calling `ull_scan_enable` instead of returning `BT_HCI_ERR_CMD_DISALLOWED`). Is this really intended?

cvinayak replied:
The change is intentional. The `BT_HCI_ERR_CMD_DISALLOWED` was not exercised due to earlier check on scanning/initiating being in disabled state, hence does not impact the sematics in practice. The new intention here is to use single connection context in two scan set (1M, Coded PHY) and it be only initialized once on the first call to ll_create_connection of 1M and/or Coded PHY from the hci.c, the later call reuses/by-passes initialization.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -47,64 +47,154 @@
 		return BT_HCI_ERR_CMD_DISALLOWED;
 	}
 
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
```
This PR seems to assume that coded phy is always supported. I think here and in other places additional guards are needed to exclude code if coded phy is not supported.

cvinayak replied:
> This PR seems to assume that coded phy is always supported.

No. The parameter phy is checked below and code short-circuits if there is not Coded PHY support in implementation.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -47,64 +47,154 @@
+	struct lll_scan *lll;
+
+	if (IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT) && is_coded_phy) {
+		lll = &scan_coded->lll;
```
It looks like this code would update the state in only one of the lll_scans, not both. Maybe factoring out the relevant code into a function and calling it for both lll_scans would be appropriate?

cvinayak replied:
The privacy support is TODO, will need to remember to address then.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1,94 +1,502 @@
+		/* C1, Tx Power is optional on the LE 1M PHY, and reserved for
+		 * for future use on the LE Coded PHY.
+		 */
+		if (lll->phy_p != BIT(2)) {
```
Please use a define instead of `BIT(2)`

cvinayak replied:
addressed.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -0,0 +1,737 @@
+			/* C1, Tx Power is optional on the LE 1M PHY, and
+			 * reserved for future use on the LE Coded PHY.
+			 */
+			if (adv->lll.phy_p != BIT(2)) {
```
Please use a define instead of `BIT(2)`

cvinayak replied:
addressed.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -0,0 +1,737 @@
+	/* No SyncInfo */
+
+	/* TODO: TxPower */
+	if (flags & BIT(6)) {
```
Please use a define/enum instead of `BIT(6)`

cvinayak replied:
addressed.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -0,0 +1,737 @@
+		si = (void *)ps;
+		si->offs = 0; /* NOTE: Filled by secondary prepare */
+		si->offs_units = 0;
+		si->interval = interval;
```
This needs to be stored as LE16.

cvinayak replied:
addressed.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1,94 +1,502 @@
+			did++;
+		}
+
+		ap->did = did;
```
As far as I can tell, this should be okay without endianness conversion. However, it would be preferable to convert endianness anyway to be able to have identical traces between LE and BE architectures.

cvinayak replied:
good catch. Fixed.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -0,0 +1,737 @@
+
+			did++;
+
+			ap->did = did;
```
As far as I can tell, this should be okay without endianness conversion. However, it would be preferable to convert endianness anyway to be able to have identical traces between LE and BE architectures.

cvinayak replied:
good catch. Fixed.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -0,0 +1,737 @@
+
+	/* TODO: AdvData */
+
+	/* Calc primary PDU len */
```
Isn't this the tertiary PDU here?

cvinayak replied:
yes, corrected.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -0,0 +1,737 @@
+						    ext_hdr_adi_adv_data);
+		pdu->len = ter_len;
+	} else {
+		t->ext_hdr_len = 0;
```
Is it really sane to just set the extended header length to 0 and thus drop the extended header?

cvinayak replied:
yes, extended header ranges from 0 to 63 octets.

cvinayak commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -467,28 +472,54 @@ endif # BT_CONN
 	int "LE Advertising Extensions Sets"
 	depends on BT_CTLR_ADV_EXT
-	default 1
+	range 1 64
```
No, just inherited the number 64 as host limited connection instances to the same 64 limit.

Action: I will change this to BT Spec. handle range, after considering implementation limitations if any (ticker id is u8_t, and LL states/roles share the 0-255 range).

cvinayak commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -214,13 +214,13 @@ enum ext_adv_aux_phy {
 	u16_t rfu:2;
 	u16_t offs_units:1;
-	u16_t sync_pkt_offs:13;
+	u16_t offs:13;
```
I will relook, if I can cant find an immediate solution, we need to revisit in the future.

cvinayak replied:
Address in the future.

cvinayak commented:
@wopu-ot Thanks for the review comments, I will address these in the coming days and hold a follow up call on resolutions taken.

cvinayak committed:
Bluetooth: controller: split: Refactor PKT_US for Adv. Ext. use

Refactor the PKT_US for reuse with Extended Scanning.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix to use advertiser random address

Fix legacy advertising started using Extended Advertising
sets to use the set advertising random address.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Initial extended scanning code

Initial extended scanning code added.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Refactor node_rx_ftr

Remove us_radio_rdy field in the footer structure as its always
a constant value which can be obtained using a functional
interface to HAL.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Move PKT_US into pdu.h

Move the definition of PKT_US into pdu.h so that it can be used
by Advertising Extensions implementations too.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix single-short ticker operation failure

Fix ticker to correctly generate the operation callback if
single-shot ticker fails to be scheduled and does not
expire.

As the collision resolution is deferred to the worker, if
a single-shot ticker is cancelled, then appropriately call
the operation callback with failure status. This is allow
the user to take necessary actions, like release resources
allocated that need to be release on the failure to schedule
the single-shot ticker expiry.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix collision resolution for single-shot ticker

Fix collision resolution to consider single-shot ticker,
current single-shot with ticks_slot as older that a next
periodic or single-shot ticker with ticks_slot.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Use uint8_t for advertising handles

Refactor to use uint8_t for advertising handles.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix extended event properties changes

Added implementation to handle updates to adv mode and event
properties of an already created advertising set.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Schedule auxiliary after primary PDU

Fix to avoid a crash due to missing aux_ptr in current
primary PDU constructed by LE Set Extended Advertising
Parameter Set, while aux_ptr was populated by LE Set
Extended Advertising Data Set in the latest double buffered
PDU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Add Extended Scanning window widening

Added implementation to calculate the window widening for
the reception of auxiliary PDU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix slave HCTO calculation

Fix missing sleep clock accuracy related jitter in the
header complete timeout calculation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Add correct aux scan ticks slot

Implemented the missing ticks_slot calculation for the
auxiliary channel PDU scanning window.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix incorrect adv_addr type

Fix incorrect adv_addr type in auxiliary PDU when address
is removed from primary and added to auxiliary PDU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix DID increment

Fix missing DID increment when AD data changed to zero len.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Minor refactor return variable

Minor refactor return variable name and fix return value
type.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix to stop preempt ticker

Fixed premature abort of events due to stale preempt ticker
expiry.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Abort auxiliary LLL if aux_ptr missing

Abort Auxiliary LLL prepare if current primary PDU is
missing the aux_ptr.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix missing enh conn complete HCI event

When Privacy is not supported in the build and Advertising
Extensions is enabled, HCI LE Enhanced Connection Complete
event was missing. This is fixed now.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Updated HCI BT_INFO of scanned PDU

Updated HCI BT_INFO of printing the parsed extended PDU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix Low Latency Prepare Done

Fix the Low Latency Prepare Done that disables ULL_LOW
context executions inside a radio event, so that
ticker_job_idle_get is invoked from ULL_LOW instead of LLL
context so that its enqueued alongwith others scheduled
in ULL_LOW. This is avoid premature disable of ULL_LOW
context executions.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix for missing Aux PDU

Fix missing Auxiliary PDU when primary PDU is calculating
the offset. Schedule the auxiliary PDU ticker before primary
PDU ticker is scheduled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Extended scanning Rx list

Added implementation to enqueue received auxiliary PDUs in
the ULL.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mtpr-ot commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -0,0 +1,269 @@
+	LL_ASSERT(!err || err == -EINPROGRESS);
+}
+
+void lll_adv_aux_offset_fill(u32_t ticks_offset, u32_t start_us,
```
Could this function be moved to ull_adv_ext to reduce code copying? And if so, could it return the aux pointer for convenience?

cvinayak replied:
yes, moved the function to ULL files.

cvinayak committed:
Bluetooth: controller: added Kconfig for max adv data length

Added a Kconfig for maximum advertising data length.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Refactor pdu.h defines

Refactor pdu.h defined in anticipation of reused in
Advertising Extensions implementation. To reuse in
calculating ticks_slots etc.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Integrate extended adv report into HCI

Integrate the generated extended adv report PDU list into
HCI layer to generate the HCI LE Extended Advertising Report
event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Change the extended scan PDU rx list

Change the implementation of extended scan PDU rx list to use
the extra pointer as the next field in the linked list node.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix compile error in periodic adv code

Fix copy-paste compile error in the implementation of the
periodic advertising.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix Ext Scanning when coded PHY unsupported

Fix implementation from starting coded PHY scan window.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix ext create conn when coded PHY unsupp

Fix implementation from performing a redundant loop checking
for coded PHY create connection parameters in HCI command.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix scan set allocation

Fix scan set allocation using BT_CTLR_SCAN_SET instead of
an incorrect duplicate BT_CTLR_SCAN_MAX.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: make ctrl_advx a BabbleSim test

Convert the ctrl_advx application into a Babblesim test.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix conditional compile of ticker ids

Fix conditional compile of ticker ids for auxiliary sets and
periodic sets.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Avoid ticker null pointer dereferencing

Avoid ticker from null pointer dereferencing when storing
operation callback functions to be called during collision
resolution.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix auxiliary channel scan window width

Fix the calculation of auxiliary channel scan window width.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix directed adv addr type returned

Fix HCI LE Directed Advertising Report event to return
correct directed advertiser's address type.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix crash due to incorrect no. of tickers

Fix segmentation fault detected by BabbleSim test due to
incorrect number of tickers being allocated.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_chan.c:
```diff
@@ -0,0 +1,40 @@
+
+#include "util/util.h"
+
+static u8_t map[5] = {0xFF, 0xFF, 0xFF, 0xFF, 0x1F};
```
Added.

cvinayak commented on subsys/bluetooth/controller/util/util.c:
```diff
@@ -55,3 +59,174 @@ int util_rand(void *buf, size_t len)
+	}
+	retry--;
+
+	util_rand((void *)&aa, sizeof(aa));
```
addressed.

cvinayak commented on subsys/bluetooth/controller/util/util.c:
```diff
@@ -55,3 +59,174 @@ int util_rand(void *buf, size_t len)
+ * - It shall have no more than eleven transitions in the least significant 16
+ *   bits.
+ */
+int util_aa_to_le32(u8_t *dst)
```
addressed as `util_aa_le32`

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_master.c:
```diff
@@ -331,7 +398,7 @@ u8_t ll_chm_update(u8_t *chm)
 	u8_t ret;
 
-	ull_conn_chan_map_set(chm);
+	ull_chan_map_set(chm);
```
will address in the future, Added a FIXME comment for now.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -772,7 +772,7 @@ u32_t radio_tmr_start_now(u8_t trx)
 	} while (now > start);
 
-	return start;
+	return start + 1;
```
The compare in the above lines is set to `start + 1`, hence this is a fix now. I consider this a minor fix and there has not been any significant impact due to BT Spec permitted active jitter of +/- 1 us.

cvinayak commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -25,23 +25,29 @@
-	TICKER_ID_ADV_LAST = ((TICKER_ID_ADV_BASE) + (BT_CTLR_ADV_MAX) - 1),
+	TICKER_ID_ADV_LAST = ((TICKER_ID_ADV_BASE) +
+			      (CONFIG_BT_CTLR_ADV_SET) - 1),
+#if (CONFIG_BT_CTLR_ADV_AUX_SET > 0)
```
Addressed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -74,9 +74,24 @@
 #if defined(CONFIG_BT_BROADCASTER)
 #define BT_ADV_TICKER_NODES ((TICKER_ID_ADV_LAST) - (TICKER_ID_ADV_STOP) + 1)
-#else
+#if defined(CONFIG_BT_CTLR_ADV_EXT) && (CONFIG_BT_CTLR_ADV_AUX_SET > 0)
```
Addressed.

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1616,6 +1633,309 @@ static void le_enh_tx_test(struct net_buf *buf, struct net_buf **evt)
+	struct bt_hci_evt_cc_status *ccst;
+	u8_t status;
+
+	status = ll_adv_aux_ad_data_set(cmd->handle, cmd->op, cmd->frag_pref,
```
Removed the guard for the implementation of this function.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1,94 +1,502 @@
+	/* Get reference to previous primary PDU data */
+	_p = (void *)&_pri->adv_ext_ind;
+	hp = (void *)_p->ext_hdr_adi_adv_data;
+	*(u8_t *)&_hp = *(u8_t *)hp;
```
I believe without `u8_t` typecast its compiler dependent to use builtin `memcpy` for structure copy.

wopu-ot replied:
Ok if you saw any issues, but putting this little trust into the compiler seems overly defensive to me.

cvinayak replied:
I stand corrected, I will fix this as this is just a struct of 1 byte (in my reply, I had confused myself with `ext_hdr_adi_adv_data` which is 254 bytes)

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1616,6 +1633,309 @@ static void le_enh_tx_test(struct net_buf *buf, struct net_buf **evt)
+	u8_t status;
+
+	set_num = cmd->set_num;
+	if (!set_num) {
```
Addressed with a FIXME comment and return of unsupported feature param val err code.

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1616,6 +1633,309 @@ static void le_enh_tx_test(struct net_buf *buf, struct net_buf **evt)
+	p = cmd->p;
+
+	phys_bitmask = BT_HCI_LE_EXT_SCAN_PHY_CODED | BT_HCI_LE_EXT_SCAN_PHY_1M;
+	do {
```
Added comments for now.

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1616,6 +1633,309 @@ static void le_enh_tx_test(struct net_buf *buf, struct net_buf **evt)
+	p = cmd->p;
+
+	phys_bitmask = BT_HCI_LE_EXT_SCAN_PHY_CODED | BT_HCI_LE_EXT_SCAN_PHY_1M;
+	do {
```
Added comments for now.

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1616,6 +1633,309 @@ static void le_enh_tx_test(struct net_buf *buf, struct net_buf **evt)
+		dup_count = -1;
+	}
+#endif
+
```
Added FIXME comment.

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1092,12 +1093,28 @@ static void le_create_connection(struct net_buf *buf, struct net_buf **evt)
+				      cmd->peer_addr.type,
+				      &cmd->peer_addr.a.val[0],
+				      cmd->own_addr_type, conn_interval_max,
+				      conn_latency, supervision_timeout, 0U);
```
I want to pass phy = 0, so that LLL classifies it as legacy scanner/initiator.

wopu-ot replied:
It seems a little odd to pass this information via the phy parameter, but ok. Could you then use a define that shows this intention (`PHY_LEGACY` or something like that)?

cvinayak replied:
Addressed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ll_addr.c:
```diff
@@ -55,9 +58,16 @@ u8_t *ll_addr_get(u8_t addr_type, u8_t *bdaddr)
+		u32_t status = ull_adv_is_enabled(0);
+
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+		if ((status & 5U) == 1U) {
```
Addressed.

cvinayak commented on tests/bluetooth/ctrl_advx/src/main.c:
```diff
@@ -0,0 +1,236 @@
+	adv_type = 0x05; /* Adv. Ext. */
+	phy_p = ADV_PHY_1M;
+	phy_s = ADV_PHY_2M;
+	err = ll_adv_params_set(handle, evt_prop, ADV_INTERVAL, adv_type,
```
Just for readability of the passed parameter (names), constants passed using defines.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -131,19 +134,30 @@ u8_t ll_adv_params_set(u16_t interval, u8_t adv_type,
 		}
+
+		/* Mark the adv set as created by extended advertising cmd */
+		adv->is_created = 3U;
```
Addressed

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1019,10 +1152,14 @@ u32_t ull_adv_is_enabled(u16_t handle)
 	}
 
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+	return BIT(0) | ((u32_t)adv->is_created << 1);
```
Addressed by using defines.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -914,13 +956,68 @@ u8_t ll_adv_enable(u8_t enable)
+			aux->interval =
+				adv->interval +
+				(HAL_TICKER_TICKS_TO_US(ULL_ADV_RANDOM_DELAY) /
+				 625U);
```
yes, its intentional, as primary PDU drifts by an amount of random delay. By having a fix value greater than the max random delay, the total auxiliary PDU count is less than or equal to total primary PDUs. This way every auxiliary PDU gets referenced by atleast one primary PDU.

wopu-ot replied:
But won't there then be also primary PDUs that point to an auxiliary PDU that will not be sent because it conflicts with another primary PDU?

cvinayak replied:
auxiliary PDU being of larger interval than its primary PDU, gets precedence in the collision resolution. The skipped primary PDU gets precedence in its next interval; in case of repeated overlap, primary and auxiliary PDUs round-robin in schedule ensuring the tuples are scheduled in air.

cvinayak committed:
Bluetooth: controller: split: Disable all adv sets unsupported

Added comments and code to make note that disabling all
advertising sets is not yet implemented.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Rework with defines for adv set states

Review rework, added comments, TODOs, FIXMEs and converted
magic number use in advertising set state flags to defines.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Set Extended Advertising feature bit

Update supported features bitmap with Extended Advertising
bit.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Remove +1 magic ticks and microseconds

Remove use of magic numbers to adjust ticks offset, replace with
vendor defines.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Move aux_offset_fill into ULL files

Move the aux pointer fill function to ULL files, so as it is
re-usable by all vendors.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix endianness in filling DID

Handle endianness when filling the DID value in advertising
PDU extended header.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Review rework use defines

Review rework updates, use defines in place of magic values.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -214,13 +214,13 @@ enum ext_adv_aux_phy {
 
 struct ext_adv_sync_info {
```
Renamed `ext_adv_*` to `pdu_adv_*` (no namespace conflict, hence no need of `ext` or `advx`)

cvinayak committed:
tests: Bluetooth: bsim_test: Include advx testing in CI

Update compile.sh and added test_script, so that
bsim_test_advx bsim_test application is built and run in
CI.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -740,6 +781,7 @@ u8_t ll_adv_enable(u8_t enable)
 #if defined(CONFIG_BT_CTLR_ADV_EXT)
 	if (pdu_adv->type == PDU_ADV_TYPE_EXT_IND) {
 		/* TBD */
+		slot_us += 1500;
```
Add TODO with what needs to be done, to calculate the PDU length in microseconds based on populated fields.

cvinayak replied:
FIXME comment has been added.

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2664,38 +3040,142 @@ static void le_advertising_report(struct pdu_data *pdu_data,
+				     struct node_rx_pdu *node_rx,
+				     struct net_buf *buf)
 {
+	const u8_t c_adv_type[] = { 0x13, 0x15, 0x10, 0xff,
+				    0x1a, /* SCAN_RSP to an ADV_SCAN_IND */
+				    0x1b, /* SCAN_RSP to an ADV_IND */
+				    0x12 };
```
Fixed.

cvinayak commented on tests/bluetooth/ctrl_advx/src/main.c:
```diff
@@ -0,0 +1,236 @@
+
+static u8_t adv_data1[] = {
+		2, BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR),
+		7, BT_DATA_NAME_COMPLETE, 'Z', 'e', 'p', 'h', 'y', 'r',
```
Explained that 2 and 7 are len values of following octets in the AD data format.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -117,10 +122,8 @@ u8_t ll_adv_params_set(u16_t interval, u8_t adv_type,
+	if (adv_type == PDU_ADV_TYPE_EXT_IND) {
 		/* legacy */
 		if (evt_prop & BIT(4)) {
 			u8_t const leg_adv_type[] = { 0x03, 0x04, 0x02, 0x00};
```
Fixed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -131,19 +134,30 @@ u8_t ll_adv_params_set(u16_t interval, u8_t adv_type,
 			 * - High duty cycle directed connectable not allowed
 			 */
 			if (((evt_prop & 0x03) == 0x03) ||
 			    ((evt_prop & 0x0C) == 0x0C)) {
```
Fixed.

cvinayak committed:
Bluetooth: controller: split: Refactor auto variable names

Review rework updates, refactor auto variable names.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix initialization of new aux header

Fix missing implementation of initialization of common
advertising header bit fields when new auxiliary channel PDU
is instantiated. This caused corrupt invalid Bluetooth
Device address being copied from uninitialised previous
auxiliary PDU instead of copying from primary channel PDU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -104,3 +538,304 @@ u8_t ll_adv_aux_set_clear(void)
+
+	/* FIXME: we are in ULL_LOW context, fill offset in LLL context */
+	pdu = lll_adv_data_curr_get(&adv->lll);
+	ull_adv_aux_lll_offset_fill(ticks_to_expire, 0, pdu);
```
Also, this needs to ensure the PDU is filled before start tick.

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -433,28 +438,68 @@ endif # BT_CONN
+	help
+	  Maximum Extended Advertising Data Length.
+
+config BT_CTLR_SCAN_AUX_SET
```
Could you give a brief explanation describing the differences between the following configurations:
* BT_CTLR_ADV_SET
* BT_CTLR_ADV_AUX_SET
* BT_CTLR_ADV_SYNC_SET
* BT_CTLR_SCAN_AUX_SET

I believe we need to make the descriptions more clear by making them match the specification better.
I would expect there to be only two configuration options:
1. BT_CTLR_ADV_SET. The maximum number of supported advertising sets. This covers legacy, extended, and periodic advertising sets
2. BT_CTLR_SYNC_SET. The maximum number of syncs to periodic advertisers.  

If this controller wants to expose more configuration options, I would propose to move them under the section for BT_LL_SW_SPLIT

cvinayak replied:
BT_CTLR_ADV_SET is the maximum number of supported advertising sets.
BT_CTLR_ADV_AUX_SET is maximum number of supported auxiliary channel sets, defines the number of actual required advertising instances on auxiliary channels. If users only need legacy advertising sets, they have the option to save a lot of memory and code space.
BT_CTLR_SCAN_AUX_SET is the maximum supported auxiliary channel scan sets, this configures maximum number of simultaneous extended advertising that is to be scanned/followed. A primary channel scan window could pick multiple extended advertisers, this Kconfig allows for how many to be scanned/followed.
BT_CTLR_ADV_SYNC_SET is equivalent to BT_CTLR_SYNC_SET.

Ok, I will move the non-HCI definable/exposable Kconfig into advanced features of BT_LL_SW_SPLIT.

rugeGerritsen reviewed (changes requested):
I would like some changes to the new and changed controller KConfig options

cvinayak commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -433,28 +438,68 @@ endif # BT_CONN
 config BT_CTLR_CHAN_SEL_2
 	bool "Channel Selection Algorithm #2"
-	depends on (BT_CONN || BT_CTLR_ADV_EXT) && BT_CTLR_CHAN_SEL_2_SUPPORT
+	depends on (BT_CONN || BT_CTLR_ADV_PERIODIC) && BT_CTLR_CHAN_SEL_2_SUPPORT
```
BT_CONN is included as per the spec requirementin v5.2, Vol6. Part B, section 4.6.12.
"A Controller that supports connections shall also support the Channel Selection Algorithm #2 feature."

When support for connection over extended advertising is added, it will enforce "select".

cvinayak commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -433,28 +438,68 @@ endif # BT_CONN
 
 config BT_CTLR_ADV_EXT
-	bool "LE Advertising Extensions"
+	bool "LE Advertising Extensions [EXPERIMENTAL]"
```
Advertising Extensions is EXPERIMENTAL in Zephyr's Bluetooth Kconfig. But yes, a conditional prompt can be included.

cvinayak commented:
> 
> 
> This seems more a sample instead of a test; consider moving to sample/bluetooth directory
> missing .yaml file?

@kruithofa as the test is now a bsim test, please re-review your change request.

cvinayak committed:
Bluetooth: controller: Re-arrange Kconfig options

Re-arrange and have conditional prompt for BT_CTLR_ADV_EXT.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -433,28 +438,50 @@ endif # BT_CONN
+config BT_CTLR_ADV_PERIODIC
+	bool "LE Periodic Advertising"
+	depends on BT_BROADCASTER
+	select BT_CTLR_CHAN_SEL_2
```
Need to add `depends on BT_CTLR_ADV_PERIODIC_SUPPORT`

rugeGerritsen reviewed (approved):
I looked through the controller Kconfig. Looks good now

cvinayak commented:
Close and re-open to retrigger CI (Gitlint, Identity/Emails, Kconfig not started for over 6 hours).

kruithofa commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c:
```diff
@@ -454,15 +540,15 @@ static int prepare(lll_is_abort_cb_t is_abort_cb, lll_abort_cb_t abort_cb,
+			err = event.curr.is_abort_cb(NULL, 0, event.curr.param,
 						     &resume_cb, &resume_prio);
-			LL_ASSERT(ret);
+			LL_ASSERT(err);
```
not really related to this PR, but I think this ASSERT should be removed; depending on how the ASSSERT is implemented this may stop the flow here.

kruithofa commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -0,0 +1,345 @@
+	trx_cnt++;
+
+	switch (lll->phy) {
+	case BIT(0):
```
Ideally we would use the PHY_1M, PHY_2M and PHY_CODED definitions from pdu.h here

kruithofa commented on subsys/bluetooth/controller/util/util.c:
```diff
@@ -55,3 +59,175 @@ int util_rand(void *buf, size_t len)
+ * - It shall have no more than eleven transitions in the least significant 16
+ *   bits.
+ */
+int util_aa_le32(uint8_t *dst)
```
It is not obvious that aa stands for access address, propose to write this out fully in a later update
